src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
         }
 
+        private void VerifyWrite()
+        {
+            TryInitializeResponseContentLength();
+
+            if (_responseContentLength.HasValue && _responseBytesWritten > _responseContentLength.Value)
+            {
+                ThrowContentLengthMismatchException();
+            }
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = int.Parse(responseHeaders.HeaderContentLength.ToString());
```

Comment:
Wow. I thought our LargeDownload test wrote over 2GB like the LargeUpload test. It looks like the LargeDownload test only has a 1MB body.

We should change the LargeDownload test to be like the LargeUpload test where we check every byte value in a few megabyte scenario, but still ensure that sending over 2GB works.